django/db/models/expressions.py

```diff
@@ -360,27 +360,30 @@ def flatten(self):
             if expr:
                 yield from expr.flatten()
 
+    @property
+    def identity(self):
+        identity = (self.__class__,)
+        output_field = self.__dict__.get('output_field')
+        if isinstance(output_field, fields.Field):
+            # Field instances need to be special cased because their equality
+            # is determined by their creation_counter.
+            name, path, args, kwargs = output_field.deconstruct()
+            args = tuple(args)
+            kwargs = tuple([
+                (key, tuple(value) if isinstance(value, list) else value)
+                for key, value in kwargs.items()
+            ])
+            identity += ((name, path, args, kwargs),)
+        else:
+            identity += (None,)
+        return identity
+
     def __eq__(self, other):
-        if self.__class__ != other.__class__:
-            return False
-        path, args, kwargs = self.deconstruct()
-        other_path, other_args, other_kwargs = other.deconstruct()
-        if (path, args) == (other_path, other_args):
-            kwargs = kwargs.copy()
-            other_kwargs = other_kwargs.copy()
-            output_field = type(kwargs.pop('output_field', None))
-            other_output_field = type(other_kwargs.pop('output_field', None))
-            if output_field == other_output_field:
-                return kwargs == other_kwargs
-        return False
+        return isinstance(other, BaseExpression) and other.identity == self.identity
 
     def __hash__(self):
-        path, args, kwargs = self.deconstruct()
-        kwargs = kwargs.copy()
-        output_field = type(kwargs.pop('output_field', None))
-        return hash((path, output_field) + args + tuple([
-            (key, tuple(value)) if isinstance(value, list) else (key, value)
-            for key, value in kwargs.items()
+        return hash(tuple([
+            tuple(part) if isinstance(part, list) else part for part in self.identity
```

Comment:
You do the `tuple(value) if isinstance(value, list) else value` dance in a couple of places. I wonder if it would be clearer to use a helper function? I'm not sure about a name, maybe `list_to_tuple`?

Reply:
The conversion from `list` to `tuple` is really hashing specific and common in other part of the code base so I guess something along `ensure_hashable` would make more sense.

Reply:
Is it worth the function call overhead?

Reply:
I don't have a strong feeling about this; I feel like this might not be necessary.